crates/sui-framework/packages/sui-framework/sources/deny_list.move

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Defines the `DenyList` type. The `DenyList` shared object is used to restrict access to
+/// instances of certain core types from being used as inputs by specified addresses in the "deny
+/// list".
+module sui::deny_list {
+    use sui::tx_context::TxContext;
+    use sui::object::{Self, UID};
+    use sui::transfer;
+    use std::vector;
+    use sui::tx_context;
+    use sui::table::{Self, Table};
+    use sui::vec_set::{Self, VecSet};
+    use sui::versioned::{Self, Versioned};
+
+    friend sui::coin;
+
+    /// Trying to create a deny list object when not called by the system address.
+    const ENotSystemAddress: u64 = 0;
+    /// The specified address to be removed is not already in the deny list.
+    const ENotDenied: u64 = 1;
+
+    #[allow(unused_const)]
+    /// The index into the deny list vector for the `sui::coin::Coin` type.
+    const COIN_INDEX: u64 = 0;
+
+
+    /// A shared object that stores the addresses that are blocked for a given core type.
+    struct DenyList has key {
+        id: UID,
+        /// the versioned deny list
+        inner: Versioned,
+    }
+
+    /// Stores the deny lists for each type
+    struct DenyListV0 has store {
+        /// A vector of lists, each element is used for a distinct core framework type
+        lists: vector<PerTypeListV0>,
+    }
+
+    /// Stores the addresses that are denied for a given core type.
+    struct PerTypeListV0 has store {
+        /// Number of object types that have been banned for a given address.
+        /// Used to quickly skip checks for most addresses
+        denied_count: Table<address, u64>,
+        /// Set of addresses that are banned for a given type.
+        /// For example with `sui::coin::Coin`: If addresses A and B are banned from using
+        /// `0x123::my_coin::MY_COIN`, this will be 0x123 -> {A, B}
```

Comment:
Did you mean `0x123::my_coin::MY_COIN` -> {A, B}
Can you also comment on whether the address needs to be canonical, and whether the bytes actually start with the bytes representation for "0x"?